resource/file/file.go

```diff
@@ -0,0 +1,566 @@
+			if !same {
+				actual.Type = TypeLink
+				status.AddDifference("hardlink", f.Destination, f.Target, "")
+				return nil
+			}
+		}
+	}
+	return err
+}
+
+// keep existing permissions if new ones are not requested
+func (f *File) diffMode(actual *File, status *resource.Status) {
+
+	if f.Type == TypeSymlink {
+		return
+	}
+
+	switch f.Mode {
+	case nil:
+		switch actual.Mode {
+		case nil: // use default perms if nothing is set
+			m := ModeType(uint32(defaultPermissions), f.Type)
+			f.Mode = &m
+			status.AddDifference("permissions", os.FileMode(0).String(), os.FileMode(*f.Mode).String(), "")
+		default:
+			f.Mode = actual.Mode
+		}
+	default: // compare to file permissions
+		switch actual.Mode {
+		case nil:
+			status.AddDifference("permissions", os.FileMode(0).String(), os.FileMode(*f.Mode).String(), "")
+		default:
+			if *actual.Mode != *f.Mode {
+				status.AddDifference("permissions", os.FileMode(*actual.Mode).String(), os.FileMode(*f.Mode).String(), "")
+			}
+		}
+	}
+}
+
+// GetFileInfo populates a File struct with data from a file on the system
+func GetFileInfo(f *File, fi os.FileInfo) error {
+	var err error
+
+	if f.State == StateUndefined {
+		f.State = StatePresent
+	}
+
+	f.Type, err = GetType(fi)
+	if err != nil {
+		return errors.Wrapf(err, "error determining type")
+	}
+
+	// set perms & higher bits for file type
+	// in POSIX permissions are not set for symlinks
+	if f.Type != TypeSymlink {
+		f.Mode = new(uint32)
+		*f.Mode = ModeType(uint32(fi.Mode()), f.Type)
+	}
+
+	// follow symlinks
+	if f.Type == TypeSymlink {
+		f.Target, err = os.Readlink(f.Destination)
+		if err != nil {
+			return errors.Wrapf(err, "error determining target of symlink")
+		}
+	}
+
+	f.UserInfo, err = UserInfo(fi)
+	if err != nil {
+		return errors.Wrapf(err, "error determining owner")
+	}
+
+	f.GroupInfo, err = GroupInfo(fi)
+	if err != nil {
+		return errors.Wrapf(err, "error determining group")
+	}
+
+	return err
+
+}
+
+// hash is used to compare file content
+func hash(b []byte) string {
+	sha := sha256.Sum256(b)
+	return hex.EncodeToString(sha[:])
+}
+
+// Delete a file Listed by the File Resource
+func (f *File) Delete() error {
+	var err error
+	_, err = os.Lstat(f.Destination)
+
+	if os.IsNotExist(err) {
+		return nil
+	}
+
+	switch f.Type {
+	case TypeDirectory:
+		err = os.RemoveAll(f.Destination)
+	default:
+		err = os.Remove(f.Destination)
+	}
+	return err
+}
+
+// Modify a file based on a file resource
+func (f *File) Modify(status *resource.Status) error {
+	var err error
+
+	if status.HasDifference("type") || status.HasDifference("hardlink") && f.action == ActionModify {
+		switch f.Force {
+		case true:
+			// if the file type changes, delete and recreate
+			err = f.Delete()
+			if err != nil {
+				return errors.Wrapf(err, "modify: unable to recreate")
+			}
+			f.action = ActionCreate
+		default:
+			return errors.New("modify: please set force=true to change file")
+		}
+	}
+
+	switch f.Type {
```

Comment:
This feels like another opportunity to factor the implementation into helper functions called inside of the `case` statements.